tools/schema_registry_tools/schemas_command_generator.py

```diff
@@ -59,21 +58,38 @@ def generate_options():
     tmp = re.sub('},$', '}]', tmp)
     j = json.loads(tmp)
 
+    # Sort by the offset
+    j.sort(key=lambda x: x['offset'])
+    # Validate if there's a duplicated offset in the offset chain
+    previous_offset = -1
+    for item in j:
+        if item['offset'] == previous_offset:
+            print(
+                f"The offset {item['offset']} is duplicated unexpectedly at key {item['key']}, hence exiting..."
+            )
+            exit(1)
+        previous_offset += 1
+
     # Generating rpk topic create commands
-    cmd_all = ''
-    cmd_all += f"#!/bin/bash\n\n"
-    for c, i in enumerate(j):
-        d = json.loads(i['key'])
-        if d['seq'] > i['offset']:
-            cmd_all = f"The seq {d['seq']} is unexpectedly higher than the \
-offset {i['offset']} at key {i['key']}. \nThat is it's likely broken, hence exiting...."
+    expected_offset = 0
+    cmd_all = "#!/usr/bin/env bash\n"
+    cmd_all += "set -euo pipefail\n\n"
+    cmd_all += "echo '"
+    for count, item in enumerate(j):
+        if item['offset'] > expected_offset:
+            gap = item['offset'] - expected_offset
+            for g in range(gap):
+                expected_offset += 1
+                cmd_all += '{"keytype":"NOOP","magic":0}, ' '\n'
 
-            break
-        cmd_all += f"echo '{i['value']}' | rpk topic produce _schemas --compression none -k '{i['key']}'\n"
-        if c == len(j) - 1:
+        if count == len(j) - 1:
+            # We're at the final record, so that piping to the rpk topic produce command
+            cmd_all += f"{item['key']}, {item['value']}\' | rpk topic produce _schemas --compression none -f '%k, %v\\n'\n"
```

Comment:
Why not run this after the loop?

Reply:
The last record should have `'` at the end and shoudn't have `\n`, thus I've separated it from the other ones.

Reply:
Oh, I see. I think the newline would be ignored by rpk.

I think it's worth moving  `| rpk topic produce _schemas --compression none -f '%k, %v\\n'\n"` outside of the loop.

Reply:
Did a quick test locally and confirmed you're right. Will sort it out then to make the logic simple.

Reply:
@twmb is this the most reasonable mechanism for sending a key and value to rpk topic produce? I.e., is there a format other than `-f '%k, %v` that you'd recommend that may be, more resilient??

Reply:
@twmb Any thoughts?

Reply:
Since this is being done programmatically, it'd be pretty easy to use size prefixing: `%K{<digits>}%k%V{<digits>}%v`, and then put the length of the key / value in brackets. Example, `%K{3}%k%v{5}%v` would parse `keyvalue` into `key` and `value`.

base64 / hex64 are also options but digits is easiest here

Reply:
Interesting! I wasn't ware of that way. Let me give it a try and see how it's easier.

Reply:
Given the `length of the key / value in brackets` applies to every entry, current implementation doesn't work since it generates a single rpk command, below for example. If the size prefixing approach is more robust, I can change the way to generate the command. WDYT @BenPope?

```
echo '{<given length of key>}, {<given length of value>}
{<given length of key>}, {<given length of value>}
{<given length of key>}, {<given length of value>}
' | rpk topic produce _schemas --compression none -f '%k, %v\n'
```

Reply:
I don't really know the answer. I can't think of a separator I'd consider safe, one that isn't possible inside a schema definition. It would have to be something not available in json (including, for example, binary encoded protobuf, which is base64 encoded). Perhaps vertical tab `0x0B` is the smallest change that should work?

If using one command per record is the only way to specify the length encoding, then let's go with that.

Reply:
rpk also supports parsing hex, so the values could be escaped with non-json characters, `\x00` for example.